R/derive_interp_records.R

```diff
@@ -44,48 +45,57 @@
 #'       TRUE ~ NA_character_
 #'     ),
 #'     # Ensure first that Age unit is "DAYS"
-#'     AGE = round(AGE * 30.4375)
-#'   ) %>%
-#'   # Interpolate the AGE by SEX
-#'   derive_interp_records(
-#'     by_vars = exprs(SEX),
-#'     parameter = "HEIGHT"
+#'     AGE = round(AGE * 30.4375),
+#'     AGEU = "DAYS"
 #'   )
 #'
-#' print(cdc_htage)
+#' # Interpolate the AGE by SEX
+#' derive_interp_records(
+#'   dataset = cdc_htage,
+#'   by_vars = exprs(SEX),
+#'   parameter = "HEIGHT"
+#' )
 derive_interp_records <- function(dataset,
                                   by_vars = NULL,
                                   parameter) {
   # Apply assertions to each argument to ensure each object is appropriate class
   assert_vars(by_vars, optional = TRUE)
   assert_character_scalar(parameter, values = c("HEIGHT", "WEIGHT", "BMI"))
-  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, AGE, L, M, S))
+  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, AGE, AGEU, L, M, S))
   if (parameter == "BMI") {
-    assert_data_frame(dataset, required_vars = exprs(!!!by_vars, AGE, L, M, S, P95, Sigma))
+    assert_data_frame(dataset, required_vars = exprs(!!!by_vars, AGE, AGEU, L, M, S, P95, Sigma))
   }
 
-  arrange(dataset, !!!by_vars, AGE)
+  # Ensure to have AGE unit in "Days"
+  ageu <- dataset %>%
+    select(AGEU) %>%
+    filter(toupper(AGEU) != "DAYS") %>%
+    group_by(AGEU) %>%
+    slice(1) %>%
+    ungroup()
 
-  # Ensure to have unique combination when by_vars is not defined
-  metadata_vars <- c("AGE", "L", "M", "S")
-  if (parameter == "BMI") {
-    metadata_vars <- append(metadata_vars, c("P95", "Sigma"))
+  if (nrow(ageu) > 0) {
+    cli_abort("The Age Unit (AGEU) from the input dataset must be in 'DAYS'")
   }
 
-  if (is.null(by_vars)) {
-    other_data_vars <- names(dataset %>% select(-all_of(metadata_vars)))
+  # Sort the data for the interpolation
+  arrange(dataset, !!!by_vars, AGE)
 
-    nb_occ <- nrow(dataset %>%
-      group_by(across(all_of(other_data_vars))) %>%
-      slice(1) %>%
-      ungroup())
+  # Ensure the uniqueness of records to interpolate
+  signal_duplicate_records(dataset, by_vars = exprs(!!!by_vars, AGE))
 
-    if (nb_occ > 1) {
-      cli_abort(paste0(
-        "The combination of ", paste(other_data_vars, collapse = ", "),
-        " must be unique. Please define `by_vars` otherwise."
-      ))
-    }
+  # Define the metadata variables to be interpolated
+  interp_vars <- c("AGE", "L", "M", "S")
+  if (parameter == "BMI") {
+    interp_vars <- append(interp_vars, c("P95", "Sigma"))
+  }
+
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
```

Comment:
```suggestion
  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars, expr(AGEU))
```
as `AGEU` should be added to `group_by()`.

Reply:
`AGEU` is the result of the difference between variables names of the `dataset` and ( `interp_vars` + `by_vars`). No need to specify it here.